### ISSUE TITLE
[SCM][SVN] Support working in a mono-repo checkout

### ISSUE THREAD

**Issue by solvingj** [conan-io/conan#3786]
When working with a local checkout of a whole SVN mono-repo, and trying to do `conan create` in some subdirectory for a component project, the SCM feature does not currently work.   Is this a common case? 
How could it work? What would be the barriers to supporting this?  This issue presents these questions. 

In our case, all our users always work with a full checkout of the SVN mono-repo, and we're implementing Conan to start working with our projects in a more modular way and breaking the dependency on the mono-repo structure over time.  However, this takes time, and so we can't use the SCM feature without support for this case. I suspect this will be extremely common for SVN and Perforce organizations.  

I think it should work because it's so common.  I don't think it's that difficult to implement, but it will likely add some complexity.  In terms of possibilities for implementation, it seems close already.  Conan can already easily deduce get all the paths segments it needs (svn root, conanfile location relative to svn root).  What more would be needed?

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by jgsogo** [comment]
You should work with SVN as you would do with Git, it is not a matter of ``conanfile.py`` being in the root of the repo or not, it is about identifying which URL is the root of your working copy.

If we want to allow the proposed workflow with SVN we do need the user to define which one is the root URL of each package: it can be set in ``scm = { 'url': '<url-to-root-of-package'}`` (no more ``auto``) or use a different variable to store this information.

**Comment by solvingj** [comment]
So, given that I want to work on these three projects: 

'https://svnserver/trunk/svn_root/the_subdirectory_with_conanfile_01'
'https://svnserver/trunk/svn_root/the_subdirectory_with_conanfile_02'
'https://svnserver/trunk/svn_root/the_subdirectory_with_conanfile_03'

I would think that I could write something like this in the conanfiles:

```
scm = {
    'type' : 'svn'
    'repo_url' : 'https://svnserver/trunk/svn_root',
    'project_path' : 'the_subdirectory_with_conanfile_01',
}
```


I don't necessarily like these variable names, but I think it demonstrates the point.  I get that this might only be relevant for SVN (but maybe also perforce?).  If this feature is going to be useful beyond GIT, I don't think the naive implementation of trying to model everything like GIT so that it doesn't get too complex is going to work out well.  They're just different.

**Comment by jgsogo** [comment]
You can use it right now with:

 - checkout root folder and define ``scm`` dict with:

    ```
    scm = {
        'type': 'svn',
        'url': 'https://svnserver/trunk/svn_root/the_subdirectory_with_conanfile_01'
    }
    ```

 - or, checkout the project folder ``svn co https://svnserver/trunk/svn_root/the_subdirectory_with_conanfile_0``, and:

    ```
    scm = {
        'type': 'svn',
        'url': 'auto'
    }
    ```
----

The solution that will fit your needs would be something like (not available now):

1. checkout the _root_ (it is not the root, as your are already using ``^/trunk/svn_root``): 

    ```
    svn co https://svnserver/trunk/svn_root
    ```

2. define the scm

    ```
    scm = {
        'type': 'svn',
        'url': 'https://svnserver/trunk/svn_root'  # <- you cannot use 'auto' because it is not the root of the repo, neither the root of the working copy.
        'project_path': 'the_subdirectory_with_conanfile_01'
    }
    ```

3. checkout will be done using ``url = scm['url'] + scm.get('project_path', '')``

----

I don't see any advantage between using the first approach over this one (in case you still want to checkout all the projects for a common ancestor directory)

**Comment by solvingj** [comment]
I'm not sure I understand exactly, but I think your number 3 at the bottom is indeed what I had envisioned if you were to add the `project_path` field.  The path would be concatenated with the appropriate "root" prefix for doing `svn checkout` and for doing local copies when it detects it's being run from within the SVN repository.   Is that how you're saying it might be able to work in the future?

**Comment by jgsogo** [comment]
I can see no advantage using third option versus first one. Why having two fields like `url` and `project_path` when you can have only `url`?

**Comment by solvingj** [comment]
Ok, reading through my own notes, I realize I was not very concise in explaining the barrier which currently exists for using SVN + SCM.  The problem is specifically related to Monorepos with many `conanfile.py` files spread out throughout the repository, one for each component. 

### Scenario 1 : User workflow:
Users work locally with a full checkout.  When running `conan create` for some component in some subdirectory of the Monorepo, the `scm` feature default behavior begins by first finding the `svn_root` and copying the whole repository to the `source` directory in cache.  This is the fundamental problem number one.  What is needed here is a way to override the default behavior, by defining a `project_path` relative to `svn_root`, and the `scm` feature should join the paths and perform the copy relative to that.  

### Scenario 2 : CI Workflow:
We have our CI building each component individually as separate jobs.  So, when being run in CI, we would want `conanfile.py` `scm` feature to checkout `svn_url` + `svn_path` instead of `svn_url` root.  

Again, these challenges are specific to the use of SVN with a Monorepo strategy, and not inherent to SVN.  While this might seem a corner case on the surface, I have heard a great many organizations that build C and C++, and use SVN and Perforce, use them as a Monorepo.  So, this might be a problem worth solving with this feature generally.  On the other hand, it might not depending on just how difficult it would be to address these two scenarios.

**Comment by solvingj** [comment]
As I have continued to think of this further, I don't think SCM feature will be able to cope with the Mono repo situation in a general way.  SCM is designed to make the decision between the two behaviors implicitly (copy from local vs scm clone/checkout).  It then has custom logic for handling each of these two cases. There are a number of assumptions made in the current design, based on the most common known cases of one-repo-per-project.  This makes sense.  While I've tried to generalize the nature of the Monorepo case for this discussion, I still don't really can't find a properly generalized condition logic, with properly generalizable behavior which I can say would make sense in a majority of cases.  

-----

So, for our Monorepo situation, we still have a problem we can't solve reasonably in the current version of Conan.  However, we might be able to solve it after https://github.com/conan-io/conan/issues/2456.


If we have a dedicated `exports()` method, we can then do the following: 

```
    def exports(self):
        if custom_condition:
           copy_files_from_local_dir()

    def source(self):
        if not custom_condition:
           get_files_from_scm()
```

**Comment by solvingj** [comment]
This might work, how will it decide whether to copy files like `exports_sources` versus doing `svn checkout`?

**Comment by jgsogo** [comment]
It will do both: clone the repo and copy the `exports_sources`.

**Comment by solvingj** [comment]
Perhaps I'm confused.  I believe SCM currently makes a decision and chooses between one of two behaviors: 

1.  SVN checkout
Or
2.  Copy files the local directory

Is this not correct?

Doing both seems redundant.

**Comment by jgsogo** [comment]
If the attributes in SCM points to the _self_ repository it will be redundant, if it points to a different repository, you may want to mix those repo with your local files. It is like the `source()` method, it is executed after the SCM [[1](https://docs.conan.io/en/latest/creating_packages/external_repo.html#recipe-and-sources-in-a-different-repo)]

**Comment by jgsogo** [comment]
I think that this issue could be related to #4249, there were some issues with some repo layouts related to the `source` step. In that PR I'm adding some logic to the auto-resolution of the remote URL, and to the local path stored in the `scm_folder.txt` file (sources will be copied from this local path).

Given that PR, the monorepo issues might be something from the past, it should work the same in the local workflow and in the cache.

**Comment by jgsogo** [comment]
Hi @solvingj, it took me a big effort to push PR #4249 forwards, it fixes an underlying bug related to the `scm_folder.txt` optimization; for SVN there was a different behavior between the local workflow and the cache one (and some slight differrences for `create` or `install`). Now I think it is fixed, at least there are a lot of tests with different repo configurations building locally and in the cache with the same results.

Now the SCM feature for SVN is working better:
 * if `url="auto"` then it will be resolved to the URL of the root of the working copy (I think it is the best we can do), and the `local_src_path` will resolve to the path of the root folder.
 * if `url="http://repo.svn/branches/project1"`, it doesn't matter the URL used to checkout the working copy, the `local_src_path` will point to the place in the filesystem corresponding to the hardcoded URL.

I hope this to fix the underlying problem within this issue, this fix will be released in v1.12 soon, I'm closing this one, but feel free to open another if the pain remains. I think it is hard, SVN is very different to Git for some use cases, but I also think we are moving in the right direction.

Thanks.
